tests/components/cast/test_media_player.py

```diff
@@ -300,6 +302,101 @@ def test_stop_discovery_called_on_stop(hass):
     assert state.state == 'unknown'
 
 
+async def test_group_media_states(hass: HomeAssistantType):
+    """Test media states are read from group if entity has no state."""
+    info = get_fake_chromecast_info()
+    full_info = attr.evolve(info, model_name='google home',
+                            friendly_name='Speaker', uuid=FakeUUID)
+
+    with patch('pychromecast.dial.get_device_status',
+               return_value=full_info):
+        chromecast, entity = await async_setup_media_player_cast(hass, info)
+
+    entity._available = True
+    entity.schedule_update_ha_state()
+    await hass.async_block_till_done()
+
+    state = hass.states.get('media_player.speaker')
+    assert state is not None
+    assert state.name == 'Speaker'
+    assert state.state == 'unknown'
+    assert entity.unique_id == full_info.uuid
+
+    group_media_status = MagicMock(images=None)
+    player_media_status = MagicMock(images=None)
+
+    # Player has no state, group is playing -> Should report 'playing'
+    group_media_status.player_is_playing = True
+    entity.multizone_new_media_status(str(FakeGroupUUID), group_media_status)
```

Comment:
To have tests be robust we should avoid calling entity methods or asserting entity attributes in tests.

Use the home assistant core to set up the platform, call services and assert state. On the other side, patch the integration library and call library registered callbacks to make the entity take action. This way we avoid interacting directly with the entity and allows us to change the entity implementation without changing tests.

Reply:
OK, the cast tests are perhaps a bit too hands on with the cast entity.
Do you have a good example of platform tests "doing it right"?

Reply:
Smartthings tests are good I think:
https://github.com/home-assistant/home-assistant/blob/71ecaa438583db5027a4344e2baf5fa55b70b6f7/tests/components/smartthings/test_switch.py